File: native-ios/WineNative/Sources/VivinoScanClient.swift
import Foundation
import UIKit

/// Scan étiquette **direct iPhone → api.vivino.com** (Bearer Keychain).
/// Contrat observé mitm : POST /v/11.0.0/scans/label + GET /v/9.1.1/vintages/{id}
/// Compression unique ici — ne pas pré-compresser en amont. Bearer jamais loggé.
enum VivinoScanClient {
    private static let base = "https://api.vivino.com"
    private static let scanPath = "/v/11.0.0/scans/label"
    private static let appVersion = "2026.29.0"
    private static let maxBytes = 480 * 1024

    /// UA proche app Vivino, OS + scale réels.
    private static var ua: String {
        let ios = UIDevice.current.systemVersion
        let scale = String(format: "%.2f", UIScreen.main.scale)
        return "Vivino regular/\(appVersion) (iPhone; iOS \(ios); Scale/\(scale))"
    }

    private static var osVersion: String {
        UIDevice.current.systemVersion
    }

    private static var phoneModel: String {
        // utsname machine (ex. iPhone15,2) si dispo, sinon générique
        var sys = utsname()
        uname(&sys)
        let mirror = Mirror(reflecting: sys.machine)
        let id = mirror.children.reduce("") { acc, el in
            guard let v = el.value as? Int8, v != 0 else { return acc }
            return acc + String(UnicodeScalar(UInt8(v)))
        }
        return id.isEmpty ? "iPhone" : id
    }

    /**
     * Compression unique pour Vivino sous maxBytes.
     * Si déjà sous la limite : pas de ré-encodage (évite double lossy).
     */
    static func compressForVivino(_ jpeg: Data) -> Data {
        if jpeg.count > 0 && jpeg.count <= maxBytes { return jpeg }
        var data = WineImageUtils.compressJPEG(jpeg, maxDimension: 1600, quality: 0.82)
        if data.count <= maxBytes { return data }
        var q: CGFloat = 0.74
        while data.count > maxBytes, q > 0.35 {
            data = WineImageUtils.compressJPEG(jpeg, maxDimension: 1400, quality: q)
            q -= 0.08
        }
        if data.count > maxBytes {
            data = WineImageUtils.compressJPEG(jpeg, maxDimension: 1100, quality: 0.45)
        }
        return data
    }

    static func labelScan(jpeg: Data) async throws -> LabelScanResult {
        guard let token = VivinoTokenStore.bearer, !token.isEmpty else {
            return LabelScanResult(
                ok: false,
                aiAvailable: false,
                aiError: "Bearer Vivino manquant",
                hint: "Admin → coller le Bearer (session app Vivino). Le scan part du téléphone, pas du serveur.",
                wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }
        let payload = compressForVivino(jpeg)
        var comps = URLComponents(string: base + scanPath)!
        var items: [URLQueryItem] = [
            .init(name: "app_version", value: appVersion),
            .init(name: "app_platform", value: "iphone"),
            .init(name: "app_phone", value: phoneModel),
            .init(name: "os_version", value: osVersion),
            .init(name: "app_caller_origin", value: "default"),
            .init(name: "language", value: "fr"),
            .init(name: "image_type", value: "jpg"),
            .init(name: "label_ocr_source", value: "vision"),
            .init(name: "add_user_vintage", value: "false"),
            .init(name: "crop_x", value: "0"),
            .init(name: "crop_y", value: "0"),
            .init(name: "crop_width", value: "1"),
            .init(name: "crop_height", value: "1"),
        ]
        if let uid = VivinoTokenStore.userId, !uid.isEmpty {
            items.append(.init(name: "user_id", value: uid))
        }
        comps.queryItems = items
        guard let url = comps.url else { throw WineAPIError.decode }

        let boundary = "Boundary+\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"image\"; filename=\"image\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(payload)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue(ua, forHTTPHeaderField: "User-Agent")
        req.setValue("*/*", forHTTPHeaderField: "Accept")
        req.setValue("fr;q=1", forHTTPHeaderField: "Accept-Language")
        req.httpBody = body
        req.timeoutInterval = 45

        let data: Data
        let resp: URLResponse
        do {
            (data, resp) = try await URLSession.shared.data(for: req)
        } catch {
            return LabelScanResult(
                ok: false, aiAvailable: false,
                aiError: "Réseau scan Vivino",
                hint: "Vérifie la connexion (Wi‑Fi / données).",
                wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if code == 401 || code == 403 {
            return LabelScanResult(
                ok: false, aiAvailable: false,
                aiError: "Token Vivino refusé (HTTP \(code))",
                hint: "Reconnecte l’app Vivino, capture un nouveau Bearer, puis Admin → Enregistrer.",
                wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }
        if code == 400 {
            let snip = String(data: data, encoding: .utf8) ?? ""
            let tooLarge = snip.localizedCaseInsensitiveContains("IMAGE_TOO_LARGE")
                || snip.localizedCaseInsensitiveContains("too large")
            return LabelScanResult(
                ok: false, aiAvailable: false,
                aiError: tooLarge ? "Image encore trop lourde pour Vivino" : "Scan Vivino 400",
                hint: tooLarge
                    ? "Recadre plus près ou photo manuelle moins nette."
                    : "Format refusé — réessaie ou saisie manuelle.",
                wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }
        guard code >= 200, code < 300,
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return LabelScanResult(
                ok: false, aiAvailable: false,
                aiError: "Réponse scan illisible (HTTP \(code))",
                hint: "Réessaie ou saisie manuelle.",
                wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }

        let matchStatus = (root["match_status"] as? String) ?? ""
        let vintageId = WineAPI.jsonInt(root["vintage_id"])
        if matchStatus.lowercased() == "matched", let vid = vintageId, vid > 0 {
            if let detail = try? await fetchVintage(vid, token: token) {
                return detail
            }
            return LabelScanResult(
                ok: true, aiAvailable: true, aiError: nil,
                hint: "Vision Vivino Matched (vintage \(vid))",
                wineName: "Vintage #\(vid)", producer: nil, wineColor: nil,
                vintage: nil, abv: nil, region: nil,
                candidates: [], vivinoQuery: nil, labelPhotoPath: nil
            )
        }
        return LabelScanResult(
            ok: true, aiAvailable: false,
            aiError: nil,
            hint: "Vision Vivino : pas de match — cherche le vin ou saisie manuelle.",
            wineName: nil, producer: nil, wineColor: nil, vintage: nil, abv: nil, region: nil,
            candidates: [], vivinoQuery: nil, labelPhotoPath: nil
        )
    }

    private static func fetchVintage(_ id: Int, token: String) async throws -> LabelScanResult {
        let url = URL(string: "\(base)/v/9.1.1/vintages/\(id)")!
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue(ua, forHTTPHeaderField: "User-Agent")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        let (data, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        // souvent public sans auth
        var payload = data
        if code >= 400 {
            var req2 = URLRequest(url: url)
            req2.setValue(ua, forHTTPHeaderField: "User-Agent")
            let (d2, _) = try await URLSession.shared.data(for: req2)
            payload = d2
        }
        guard let root = try? JSONSerialization.jsonObject(with: payload) as? [String: Any] else {
            throw WineAPIError.decode
        }
        let wine = root["wine"] as? [String: Any] ?? [:]
        let winery = wine["winery"] as? [String: Any] ?? [:]
        let region = wine["region"] as? [String: Any] ?? [:]
        let name = (wine["name"] as? String) ?? (root["name"] as? String)
        let producer = winery["name"] as? String
        let regionName = region["name"] as? String
        let wineId = WineAPI.jsonInt(wine["id"]) ?? 0
        let year = WineAPI.jsonInt(root["year"])
        let typeId = WineAPI.jsonInt(wine["type_id"])
        let color: String? = {
            switch typeId {
            case 1: return "rouge"
            case 2: return "blanc"
            case 3: return "effervescent"
            case 4: return "rose"
            case 7: return "fortifie"
            case 24: return "orange"
            default: return nil
            }
        }()
        let stats = root["statistics"] as? [String: Any]
        let rating = WineAPI.jsonDouble(stats?["ratings_average"])
        var photo: String?
        if let img = root["image"] as? [String: Any] {
            if let loc = img["location"] as? String {
                photo = loc.hasPrefix("//") ? "https:" + loc : loc
            } else if let vars = img["variations"] as? [String: Any],
                      let m = vars["medium"] as? String {
                photo = m.hasPrefix("//") ? "https:" + m : m
            }
        }
        let hit = VivinoHit(
            bid: wineId,
            wineName: name ?? "Vin",
            producer: producer,
            styleFr: color,
            photoURL: photo,
            vintage: year,
            country: nil,
            region: regionName,
            vivinoRating: rating,
            vivinoURL: wineId > 0 ? "https://www.vivino.com/wines/\(wineId)" : nil
        )
        return LabelScanResult(
            ok: true,
            aiAvailable: true,
            aiError: nil,
            hint: "Vision Vivino (scan iPhone) — \(matchHint(name, producer))",
            wineName: name,
            producer: producer,
            wineColor: color,
            vintage: year,
            abv: nil,
            region: regionName,
            candidates: [hit],
            vivinoQuery: [producer, name].compactMap { $0 }.joined(separator: " "),
            labelPhotoPath: nil
        )
    }

    private static func matchHint(_ name: String?, _ producer: String?) -> String {
        [producer, name].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: " · ")
    }
}

File: native-android/app/src/main/java/fr/eiter/plexiwine/VivinoScanClient.kt
package fr.eiter.plexiwine

import android.content.Context
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Scan étiquette **direct téléphone → api.vivino.com**.
 * Contrat mitm : POST /v/11.0.0/scans/label + GET vintages/{id}
 *
 * Compression unique ici (limite API ~0.5 Mo). Ne pas pré-compresser en amont.
 * Bearer jamais loggé.
 */
object VivinoScanClient {
    private const val BASE = "https://api.vivino.com"
    private const val SCAN = "/v/11.0.0/scans/label"
    private const val APP_VERSION = "2026.29.0"
    /** Limite soft sous le plafond Vivino IMAGE_TOO_LARGE (0.5 Mo). */
    private const val MAX_BYTES = 480 * 1024

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .build()

    /** UA proche app Vivino, version OS + modèle réels. */
    private fun userAgent(): String {
        val ver = Build.VERSION.RELEASE?.ifBlank { null } ?: "14"
        val model = Build.MODEL?.replace(Regex("\\s+"), "")?.ifBlank { null } ?: "Android"
        return "Vivino regular/$APP_VERSION (Linux; Android $ver; $model)"
    }

    private fun queryParams(uid: String?): String {
        val ver = Build.VERSION.RELEASE?.ifBlank { null } ?: "14"
        val model = Build.MODEL?.replace(Regex("\\s+"), "")?.ifBlank { null } ?: "Android"
        return buildString {
            append("app_version=$APP_VERSION")
            append("&app_platform=android")
            append("&app_phone=").append(model)
            append("&os_version=").append(ver)
            append("&app_caller_origin=default&language=fr&image_type=jpg&label_ocr_source=vision")
            append("&add_user_vintage=false&crop_x=0&crop_y=0&crop_width=1&crop_height=1")
            if (!uid.isNullOrEmpty()) append("&user_id=").append(uid)
        }
    }

    /**
     * Compression unique pour Vivino : une seule passe JPEG sous [MAX_BYTES].
     * Si l'entrée est déjà sous la limite, on ne ré-encode pas (évite double lossy).
     */
    private fun compressForVivino(jpeg: ByteArray): ByteArray {
        if (jpeg.size in 1..MAX_BYTES) return jpeg
        var data = ImageUtils.compressJPEG(jpeg, maxDimension = 1600, quality = 82)
        if (data.size <= MAX_BYTES) return data
        var q = 74
        while (data.size > MAX_BYTES && q > 35) {
            data = ImageUtils.compressJPEG(jpeg, maxDimension = 1400, quality = q)
            q -= 8
        }
        if (data.size > MAX_BYTES) {
            data = ImageUtils.compressJPEG(jpeg, maxDimension = 1100, quality = 45)
        }
        return data
    }

    suspend fun labelScan(ctx: Context, jpeg: ByteArray): LabelScanResult = withContext(Dispatchers.IO) {
        val token = VivinoTokenStore.bearer(ctx)
        if (token.isNullOrEmpty()) {
            return@withContext LabelScanResult(
                ok = false,
                aiAvailable = false,
                aiError = "Bearer Vivino manquant",
                hint = "Admin → coller le Bearer (session app Vivino). Le scan part du téléphone, pas du serveur."
            )
        }
        val payload = compressForVivino(jpeg)
        val uid = VivinoTokenStore.userId(ctx)
        val ua = userAgent()
        val body = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart(
                "image", "image",
                payload.toRequestBody("image/jpeg".toMediaType())
            )
            .build()
        val req = Request.Builder()
            .url("$BASE$SCAN?${queryParams(uid)}")
            .post(body)
            .header("Authorization", "Bearer $token")
            .header("User-Agent", ua)
            .header("Accept", "*/*")
            .header("Accept-Language", "fr;q=1")
            .build()
        try {
            client.newCall(req).execute().use { resp ->
                val code = resp.code
                val text = resp.body?.string().orEmpty()
                if (code == 401 || code == 403) {
                    return@withContext LabelScanResult(
                        ok = false, aiAvailable = false,
                        aiError = "Token Vivino refusé (HTTP $code)",
                        hint = "Reconnecte l’app Vivino, capture un nouveau Bearer, puis Admin → Enregistrer."
                    )
                }
                if (code == 400) {
                    val snip = text.take(120)
                    val tooLarge = snip.contains("IMAGE_TOO_LARGE", ignoreCase = true)
                        || snip.contains("too large", ignoreCase = true)
                    return@withContext LabelScanResult(
                        ok = false, aiAvailable = false,
                        aiError = if (tooLarge) "Image encore trop lourde pour Vivino" else "Scan Vivino 400",
                        hint = if (tooLarge) "Recadre plus près ou photo manuelle moins nette."
                        else "Format refusé — réessaie ou saisie manuelle."
                    )
                }
                if (code !in 200..299) {
                    return@withContext LabelScanResult(
                        ok = false, aiAvailable = false,
                        aiError = "Scan Vivino HTTP $code",
                        hint = "Réseau ou API Vivino — réessaie dans un instant."
                    )
                }
                val root = JSONObject(text)
                val matchStatus = root.optString("match_status", "")
                val vintageId = if (root.isNull("vintage_id")) null else root.optInt("vintage_id")
                if (matchStatus.equals("Matched", true) && vintageId != null && vintageId > 0) {
                    return@withContext fetchVintage(token, vintageId, ua)
                }
                return@withContext LabelScanResult(
                    ok = true,
                    aiAvailable = false,
                    hint = "Vision Vivino : pas de match — cherche le vin ou saisie manuelle."
                )
            }
        } catch (e: Exception) {
            return@withContext LabelScanResult(
                ok = false, aiAvailable = false,
                aiError = "Réseau scan Vivino",
                hint = e.message?.take(80) ?: "Vérifie la connexion (Wi‑Fi / données)."
            )
        }
    }

    private fun fetchVintage(token: String, vintageId: Int, ua: String): LabelScanResult {
        val url = "$BASE/v/9.1.1/vintages/$vintageId"
        fun get(auth: Boolean): JSONObject? {
            val b = Request.Builder().url(url).get().header("User-Agent", ua).header("Accept", "application/json")
            if (auth) b.header("Authorization", "Bearer $token")
            return try {
                client.newCall(b.build()).execute().use { r ->
                    if (r.code !in 200..299) return null
                    JSONObject(r.body?.string().orEmpty())
                }
            } catch (_: Exception) {
                null
            }
        }
        val root = get(true) ?: get(false)
            ?: return LabelScanResult(
                ok = true, aiAvailable = true,
                hint = "Matched vintage $vintageId (détail indisponible)",
                wineName = "Vintage #$vintageId"
            )
        val wine = root.optJSONObject("wine") ?: JSONObject()
        val winery = wine.optJSONObject("winery") ?: JSONObject()
        val region = wine.optJSONObject("region") ?: JSONObject()
        val name = wine.optString("name").ifBlank { root.optString("name") }
        val producer = winery.optString("name").ifBlank { null }
        val regionName = region.optString("name").ifBlank { null }
        val wineId = wine.optInt("id", 0)
        val year = root.optString("year").toIntOrNull()
        val typeId = if (wine.isNull("type_id")) null else wine.optInt("type_id")
        val color = when (typeId) {
            1 -> "rouge"
            2 -> "blanc"
            3 -> "effervescent"
            4 -> "rose"
            7 -> "fortifie"
            24 -> "orange"
            else -> null
        }
        val stats = root.optJSONObject("statistics")
        val rating = stats?.optDouble("ratings_average")?.takeIf { !it.isNaN() }
        var photo: String? = null
        root.optJSONObject("image")?.let { img ->
            val loc = img.optString("location")
            if (loc.isNotBlank()) {
                photo = if (loc.startsWith("//")) "https:$loc" else loc
            } else {
                img.optJSONObject("variations")?.optString("medium")?.let { m ->
                    if (m.isNotBlank()) photo = if (m.startsWith("//")) "https:$m" else m
                }
            }
        }
        val hit = VivinoHit(
            bid = wineId,
            wineName = name.ifBlank { "Vin" },
            producer = producer,
            styleFr = color,
            photoURL = photo,
            vintage = year,
            region = regionName,
            vivinoRating = rating,
            vivinoURL = if (wineId > 0) "https://www.vivino.com/wines/$wineId" else null
        )
        return LabelScanResult(
            ok = true,
            aiAvailable = true,
            hint = "Vision Vivino (scan téléphone) — ${listOfNotNull(producer, name).joinToString(" · ")}",
            wineName = name.ifBlank { null },
            producer = producer,
            wineColor = color,
            vintage = year,
            region = regionName,
            candidates = listOf(hit),
            vivinoQuery = listOfNotNull(producer, name).joinToString(" ")
        )
    }
}
